Task3/src/Services/Customer.hs

```diff
@@ -1,34 +1,36 @@
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
 {-# LANGUAGE TypeApplications #-}
 
 module Services.Customer
     ( getCustomers
     , getCustomer
     , addCustomer
     , editCustomer
-    , deleteCustomer) where
+    , deleteCustomer
+    , searchCustomers) where
 
 import Data.Models (CustomerModel)
-import Data.Entities (Customer (Customer))
+import Data.Entities (Customer (..))
 import qualified Repositories.OrderRepository as OrderRep
 import qualified Repositories.ProductRepository as ProductRep
 import Repositories.GenericRepository.GenericRepository
 import Mappings.Mappings (mapCustomerToModel, mapModelToCutomer)
-
+import Data.SearchModels (CustomerSearchModel(..))
+import Repositories.FilterApplier (applyFilter)
+import Data.List (isInfixOf)
 
 getCustomers :: IO [CustomerModel]
-getCustomers = do
-    csts <- getList
-    return (map (\o -> mapCustomerToModel o Nothing Nothing) csts)
+getCustomers = map (\o -> mapCustomerToModel o Nothing Nothing) <$> getList
 
 getCustomer :: Int -> IO (Maybe CustomerModel)
-getCustomer custId = do
-    customer <- get custId
-    case customer of
-        Nothing -> return Nothing
-        Just value -> do
-            orders <- OrderRep.getOrdersByCustomerId custId
-            prodWithIds <- ProductRep.getProductsWithOrdersId
-            return (Just . mapCustomerToModel value (Just orders) . Just $ prodWithIds)
+getCustomer custId =
+    get custId >>= getCustomerModel
+    where
+        getCustomerModel Nothing = return Nothing
+        getCustomerModel (Just value) =
+                let orders = OrderRep.getOrdersByCustomerId custId
+                    prodWithIds = ProductRep.getProductsWithOrdersId
+                in (\ord prodId -> Just . mapCustomerToModel value (Just ord) . Just $ prodId) <$> orders <*> prodWithIds
```

Comment:
👍 that's the place I was talking about (where you naturally may need applicative)